rq/queue.py

```diff
@@ -252,33 +252,35 @@ def enqueue_call(self, func, args=None, kwargs=None, timeout=None,
             depends_on=depends_on, timeout=timeout, id=job_id,
             origin=self.name, meta=meta)
 
-        # If job depends on an unfinished job, register itself on it's
-        # parent's dependents instead of enqueueing it.
-        # If WatchError is raised in the process, that means something else is
-        # modifying the dependency. In this case we simply retry
+        # If a _dependent_ job depends on any unfinished job, register all the
+        #_dependent_ job's dependencies instead of enqueueing it.
+        #
+        # `Job#fetch_dependencies` sets WATCH on all dependencies. If
+        # WatchError is raised in the when the pipeline is executed, that means
+        # something else has modified either the set of dependencies or the
+        # status of one of them. In this case, we simply retry.
         if depends_on is not None:
-            if not isinstance(depends_on, self.job_class):
-                depends_on = self.job_class(id=depends_on,
-                                            connection=self.connection)
             with self.connection.pipeline() as pipe:
                 while True:
                     try:
-                        pipe.watch(depends_on.key)
-
-                        # If the dependency does not exist, raise an
-                        # exception to avoid creating an orphaned job.
-                        if not self.job_class.exists(depends_on.id,
-                                                     self.connection):
-                            raise InvalidJobDependency('Job {0} does not exist'.format(depends_on.id))
-
-                        if depends_on.get_status() != JobStatus.FINISHED:
-                            pipe.multi()
-                            job.set_status(JobStatus.DEFERRED)
-                            job.register_dependency(pipeline=pipe)
-                            job.save(pipeline=pipe)
-                            job.cleanup(ttl=job.ttl, pipeline=pipe)
-                            pipe.execute()
-                            return job
+                        dependencies = job.fetch_dependencies(
+                            watch=True,
+                            pipeline=pipe
+                        )
+
+                        pipe.watch(job.dependencies_key)
+
+                        pipe.multi()
+
+                        for dependency in dependencies:
+                            if dependency.get_status() != JobStatus.FINISHED:
```

Comment:
`dependency.get_status()` makes a round trip to Redis just to get the job's status. We should check `dependency._status` instead.

Reply:
In 6f685b8b01945feea624bab98aa99ed01e467cdf

Reply:
In 58edab45ccfb042ae6a61a119901ce2dadddfbdd - I have updated `get_status`, adding a `refresh` parameter: 

https://github.com/rq/rq/blob/58edab45ccfb042ae6a61a119901ce2dadddfbdd/rq/job.py#L143

This feels a little bit like interface bloat, but also let's us avoid exposure of `Job`'s private `_status` attribute:

https://github.com/rq/rq/blob/58edab45ccfb042ae6a61a119901ce2dadddfbdd/rq/queue.py#L277

Reply:
`get_status(refresh=True)` is a wonderful idea 👍